File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        // time:O(n) and space:O(n)
        Map<Integer,Integer> mp = new HashMap<>();
        for(int i = 0;i<nums.length;i++){
            if(mp.containsKey(nums[i])){
                return nums[i];
            }
            else{
                mp.put(nums[i],1);
            }
        }
        return -1;
    }
}

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        for(auto num:nums)
        {
            int cur = abs(num);
            if(nums[cur] < 0){
                return cur;
            }
            else{
                nums[cur] *= -1;
            }
        }
        return -1;
    }
};